### PR TITLE
Fixing edge cases in picture id munging.

### PR DESCRIPTION
**Pull request by boks1971**
Changes

1. Check the RTP sequence number order before VP8 temporal layer
   filtering and use that ordering result while doing temporal
   layer filtering.

    In a sequence like below
       o Packet 10 -> Picture ID 10
       o Packet 11 -> missing
       o Packet 12 -> Picture ID 11
     it is not known if packet 11 will belong to Picture ID 10 or
     Picture ID 11. The problem becomes a lot more tricky if there
     is a burst loss and there is a larger hole in the picture id
     space also as a result.

     So, in the event of a packet loss, forward even if the current
     packet belongs to a layer that can be dropped. More comments
     in code.

2. Use result of sequence number ordering check while doing VP8 picture id munging.

3. When adding to missing picture id cache, have to include picture ids including
   both ends. As a picture can span multiple packets and it is not known which
   picture the packet belongs to, have to include both ends also in missing
   picture id cache in the event of a gap.

4. As a picture can span multiple packets, it is not possible to have a simple
   map of missing picture ids as an entry cannot be deleted if an out-of-order
   picture id is received. There may be more missing packets belonging to that
   picture id that is yet to be received.

   So, have to use an ordered map and truncate the map if it grows too large.

   Picked this for ordered map - https://github.com/elliotchance/orderedmap.
   Has a simple API, had the highest number of stars of all the ones I checked.
   And there are benchmarks.
   The author also wrote a medium post at https://medium.com/swlh/an-ordered-map-in-go-436634692381

   Another one which I looked at is - https://github.com/wk8/go-ordered-map.
   The author of that wrote at https://morioh.com/p/990229f32171 and has a
   bunch of other options at the end of that post (but does not include the
   one I picked above). None of those have that many stars.

Testing:
--------
- Set max temporal layers to 0 so that temporal filtering happens and run for
an hour on sample app.

### CODE REVIEW

File: pkg/sfu/downtrack.go [hunk 1]
```diff
@@ -30,6 +32,15 @@ const (
 	RTPBlankFramesMax             = 6
 )
 
+type SequenceNumberOrdering int
+
+const (
+	SequenceNumberOrdering_CONTIGUOUS = iota
```
Comment (cnderrauber) [inline pkg/sfu/downtrack.go:38]: set const type, and camelcase style
SequenceNumberOrderingContiguous SequenceNumberOrdering = iota

File: pkg/sfu/downtrack.go [hunk 2]
```diff
@@ -1470,7 +1483,7 @@ func (m *Munger) UpdateAndGetSnTs(extPkt *buffer.ExtPacket) (uint16, uint32, err
 	m.lastTS = mungedTS
 	m.lastMarker = extPkt.Packet.Marker
 
-	return mungedSN, mungedTS, nil
+	return mungedSN, mungedTS, SequenceNumberOrdering(ordering), nil
```
Comment (cnderrauber) [inline pkg/sfu/downtrack.go:1486]: set const type to SequenceNumberOrdering, then here not need the cast

File: pkg/sfu/downtrack.go [hunk 3]
```diff
@@ -1625,25 +1652,58 @@ func (v *VP8Munger) UpdateAndGet(extPkt *buffer.ExtPacket, maxTemporalLayer int3
...
+			return nil, ErrFilteredVP8TemporalLayer
+		} else {
+			for lostPictureId := prevMaxPictureId; lostPictureId <= extPictureId; lostPictureId++ {
+				v.missingPictureIds.Set(lostPictureId, v.pictureIdOffset)
+			}
+
+			// trim cache if necessary
+			for v.missingPictureIds.Len() > 50 {
+				el := v.missingPictureIds.Front()
+				key, ok := el.Key.(int32)
```
Comment (cnderrauber) [inline pkg/sfu/downtrack.go:1686]: I think the check and type cast is not need, you always set int32 keys.

**Review by cnderrauber — APPROVED** [review]

File: pkg/sfu/downtrack.go (hunk 1 above)
Comment (boks1971) [reply pkg/sfu/downtrack.go:38]: @cnderrauber Is that the preferred style in golang? Then I should change my signal/state definitions in `StreamAllocator` too.

Comment (cnderrauber) [reply pkg/sfu/downtrack.go]: Yes, it is golang naming convention

Comment (boks1971) [reply pkg/sfu/downtrack.go]: Got it. Will change here. Will change the `StreamAllocator` stuff later.

**Comment by boks1971** [comment]
@cnderrauber I have made these two commits to address your comments
- https://github.com/livekit/livekit-server/pull/180/commits/1784f42ee9eff56c08edaee6a7107870530f17c3
- https://github.com/livekit/livekit-server/pull/180/commits/5bf95936f42e8281acd8eae5f37e00d2474e6535

Can you please have a quick look when you get a chance and let me know if these address your comments?

**Comment by cnderrauber** [comment]
totally addressed. nice work

**Comment by boks1971** [comment]
Thank you @cnderrauber .

@davidzhao Going to merge on the back of Jie's review, but if you have comments, please send them my way and I will address it in a subsequent PR.

**Pull request merged**